Framework/PythonInterface/mantid/geometry/src/Exports/CSGObject.cpp

```diff
@@ -5,18 +5,50 @@
 //   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
 // SPDX - License - Identifier: GPL - 3.0 +
 #include "MantidGeometry/Objects/CSGObject.h"
+#include "MantidGeometry/Rendering/GeometryTriangulator.h"
+#include "MantidPythonInterface/core/Converters/NDArrayTypeIndex.h"
+#include "MantidPythonInterface/core/Converters/WrapWithNDArray.h"
 #include "MantidPythonInterface/core/GetPointer.h"
+
 #include <boost/python/class.hpp>
 #include <boost/python/copy_const_reference.hpp>
 #include <boost/python/register_ptr_to_python.hpp>
+#include <boost/python/self.hpp>
+#include <boost/python/suite/indexing/vector_indexing_suite.hpp>
+
+#include <numpy/arrayobject.h>
+
+#define PY_ARRAY_UNIQUE_SYMBOL GEOMETRY_ARRAY_API
+#define NO_IMPORT_ARRAY
 
+using Mantid::Geometry::IObject;
+using namespace Mantid::PythonInterface::Converters;
 using Mantid::Geometry::BoundingBox;
 using Mantid::Geometry::CSGObject;
-using Mantid::Geometry::IObject;
+using Mantid::Geometry::detail::GeometryTriangulator;
 using namespace boost::python;
 
 GET_POINTER_SPECIALIZATION(CSGObject)
 
+boost::python::object wrapMeshWithNDArray(const CSGObject *self) {
+  // PyArray_SimpleNewFromData doesn't interact well with smart pointers so use raw pointer
+
+  auto localTriangulator = new GeometryTriangulator(self);
+  auto vertices = localTriangulator->getTriangleVertices();
+  auto triangles = localTriangulator->getTriangleFaces();
+  size_t numberTriangles = localTriangulator->numTriangleFaces();
+  npy_intp dims[3] = {static_cast<int>(numberTriangles), 3, 3};
+  auto *meshCoords = new double[dims[0] * dims[1] * dims[2]];
+  for (size_t corner_index = 0; corner_index < triangles.size(); ++corner_index) {
+    for (size_t xyz = 0; xyz < 3; xyz++) {
+      meshCoords[3 * corner_index + xyz] = vertices[3 * triangles[corner_index] + xyz];
+    } // for each coordinate of that corner
+  }   // for each corner of the triangle
+
+  PyObject *ndarray = Impl::wrapWithNDArray(meshCoords, 3, dims, NumpyWrapMode::ReadOnly, OwnershipMode::Python);
```

Comment:
As this is a copy do we need to set the read-only flag? I'm not sure it's what users would expect?